Archive/Imo/Imo1998Q3.lean

```diff
@@ -0,0 +1,622 @@
+  simp only [d]
+  rw [Nat.divisors_prime_pow hp]
+  rw [Finset.card_map]
+  simp
+
+/--
+Proof that `dComputed n = d n`, using the fact that `d` is multiplicative
+and `d(pᵏ) = k + 1` for `p` prime.
+-/
+lemma d_eq_dComputed (n : ℕ) : d n = dComputed n := by
+  cases n
+  · rfl
+  exact Nat.card_divisors <| by lia
+
+/-- Proof that `d(n²) = ∏ᵢ(2kᵢ + 1)` for `n = ∏ᵢpᵢᵏᵢ` (prime factorization). -/
+lemma d_n2 (n : ℕ) (hn : n ≠ 0) : d (n ^ 2) = n.factorization.prod (fun _ k ↦ 2 * k + 1) := by
+  rw [d, Nat.card_divisors <| pow_ne_zero 2 hn, Nat.factorization_pow,
+    Nat.primeFactors_pow _ two_ne_zero]
+  rfl
+
+/-- Proof that if every number in a list is odd, the product of the list is odd. -/
+lemma list_prod_odd (l : List ℕ) : (∀ x ∈ l, Odd x) → Odd l.prod := by
+  induction l <;> grind [Odd.mul]
+
+/-- We begin by first proving that every value of `k` must be an odd number. -/
+theorem k_is_odd (k : ℕ)
+    (h : ∃ n : ℕ, n ≠ 0 ∧ d (n ^ 2) = k * d n) :
+    (k % 2 = 1) := by
+  obtain ⟨n, hn⟩ := h
+  let nfac := n.factorization
+  have hprime : ∀ p ∈ nfac.support, Nat.Prime p := by
+    intro p hp
+    dsimp [nfac] at hp
+    exact Nat.prime_of_mem_primeFactors hp
+  have hnfac := Nat.eq_factorization_iff hn.left hprime
+  have hneqnfac : (nfac.prod fun x1 x2 ↦ x1 ^ x2) = n := hnfac.mp (by rfl)
+  have hdn2odd : (d (n ^ 2)) % 2 = 1 := by
+    have hn2eqnfac : (nfac.prod fun p k ↦ p ^ (2 * k)) = n ^ 2 := calc
+      (nfac.prod fun p k ↦ p ^ (2 * k))
+      _ = (nfac.prod fun p k ↦ (p ^ k) ^ 2) := by
+          apply Finset.prod_congr rfl
+          intro p _
+          simp only
+          rw [pow_mul']
+      _ = (nfac.prod fun p k ↦ p ^ k) ^ 2 := by
+          unfold Finsupp.prod
+          simp only
+          rw [Finset.prod_pow]
+      _ = n ^ 2 := by rw [hneqnfac]
+    have hd : d (n ^ 2) = (nfac.prod fun _ x2 ↦ ((2 * x2) + 1)) := d_n2 n hn.left
+    let powers := nfac.support.toList.map (fun x ↦ 2 * nfac x + 1)
+    have hallodd : (∀ i ∈ powers, Odd i) := by
+      intro i hi
+      rw [Nat.odd_iff]
+      rw [List.mem_map] at hi
+      obtain ⟨p, _, rfl⟩ := hi
+      simp
+    have hoddmul : (Odd (nfac.prod fun _ x2 ↦ ((2 * x2) + 1))) := by
+      have h_eq : (nfac.prod fun _ x2 ↦ ((2 * x2) + 1)) = powers.prod := by
+        dsimp [powers]
+        rw [Finsupp.prod]
+        rw [Finset.prod_eq_multiset_prod]
+        simp
+      rw [h_eq]
+      exact list_prod_odd powers hallodd
+    rw [hd]
+    exact Nat.odd_iff.mp hoddmul
+  rw [hn.right] at hdn2odd
+  have hodd : Odd (k * (d n)) := Nat.odd_iff.mpr hdn2odd
+  have hkodd : Odd k := (Nat.odd_mul.mp hodd).left
+  exact Nat.odd_iff.mp hkodd
+
+/-- Proof that we can pick `t` new and distinct primes given a list of `t` primes. -/
+lemma exists_distinct_primes (t n : ℕ) (hnneq0 : n ≠ 0) (primes : List ℕ) :
+    (∃ l : List ℕ, (∀ i ∈ l, Nat.Prime i ∧ i ∉ primes ∧ Nat.Coprime i n)
+        ∧ List.length l = t ∧ List.Nodup l) := by
```

Comment:
This has nothing to do with primes. It should be a general statement about infinite subsets of a type.